cmd/tests/doc.go

```diff
@@ -0,0 +1,8 @@
+// Package tests contains integration tests that run k6 commands, and interact
+// with standard I/O streams. They're the highest level tests we have, just
+// below E2E tests that execute the k6 binary. Since they initialize all
+// internal k6 components similarly to how a user would, they're very useful,
+// but also very expensive to run. They're also brittle, as they depend on large
+// parts of the codebase. When in doubt, prefer adding lower-level unit tests
+// first, and an integration test only if necessary.
```

Comment:
The APIs certainly won't change much, but since they test a large portions of the system, integration tests are brittle by definition. Any issue in the underlying packages might trigger a failure in _all_ integration tests that depend on it. The failure matrix is much larger with integration tests than for unit tests.

In addition to this, integration tests are _much_ more expensive to run. I really don't want us to rely solely on them for any new features, since I would prefer to run them only occasionally locally (say, before `git push`). Whereas I can run a batch of unit tests much more quickly, and rely on them as a more frequent check while developing.

Related to this, prefering integration tests to unit tests goes against the [traditional test pyramid](https://martinfowler.com/articles/practical-test-pyramid.html). I actually don't think we need to prefer one over the other at all. All levels of testing complement each other, but it's clear that we should have _mostly_ unit tests, _many_ integration tests, and _some_ E2E tests.

The fact some of our unit tests are unreliable and flaky is not a reason to avoid them, but a reason to make them more reliable. Having to refactor them more frequently is the price of admission, and not a reason to prefer higher-level tests.

Reply:
> The APIs certainly won't change much, but since they test a large portions of the system, integration tests are brittle by definition. Any issue in the underlying packages might trigger a failure in all integration tests that depend on it. The failure matrix is much larger with integration tests than for unit tests.

No, it won't, these integration tests don't actually use most of the underlying packages directly :sweat_smile: If you can compile the `k6` binary, you should pretty much be able to run the integration tests...

Again, these tests basically rely only on the user-facing APIs, which we try very, very hard to keep somewhat stable. They are the very opposite of brittle.

> I actually don't think we need to prefer one over the other at all. All levels of testing complement each other, but it's clear that we should have mostly unit tests, many integration tests, and some E2E tests.
> 
> The fact some of our unit tests are unreliable and flaky is not a reason to avoid them, but a reason to make them more reliable. Having to refactor them more frequently is the price of admission, and not a reason to prefer higher-level tests.

Mostly agree with this, though with a clarification. Low-level **unit** tests that verify a single API are great. High-level integration tests like these that verify the whole application (basically from the user's perspective) are also great. The more you go towards the middle, the worse things get. 

The most horrible and flaky tests we have are these mid-level tests that first have to do a whole lot of setup and initialization of a whole bunch of components and mocks. And then they have to inject and verify very complicated things. Basically, the old `Engine` and `ExecutionScheduler` tests, a bunch of the tests around the executors and `js/`. Most of them would be better off as integration tests, since they try to set up the whole world before the test can be executed, and they do it by somewhat replicating what happens during a normal k6 run, but not quite exactly the same. So, any API change breaks them, and they often are not very realistic or good tests, i.e. they are the brittle ones.

Reply:
> No, it won't, these integration tests don't actually use most of the underlying packages directly :sweat_smile: If you can compile the k6 binary, you should pretty much be able to run the integration tests...

> Again, these tests basically rely only on the user-facing APIs, which we try very, very hard to keep somewhat stable. They are the very opposite of brittle.

I think we have different definitions of "brittle" :smile:

Having stable APIs doesn't matter if, e.g. `cmdRun.run()` returns an error, which can happen _for any reason_ in _any of the underlying packages_. The `k6` binary will build just fine, but the error will happen at runtime. You can try it yourself, just return a fake error from `cmdRun.run()` and see all integration tests fail. This is a contrived example, but a failure could occur deep within some dependent module, and trigger similar broad failures.

This is the opposite of a unit test where the only reason it will fail (or _should_ fail, assuming it's not flaky) is if the unit under test fails. It's not dependent on any other part of the system, and so, by definition, unit test are much more stable than higher level tests.

The reason you want to have both is because the integration test failure often doesn't tell you anything about the root cause. Preferably you'd want to have at least one unit test failure that would pinpoint exactly the root cause, and then you wouldn't really spend time debugging integration test failures at all. This is why it's important to have both. Lower level tests to debug and fix things quickly. And higher level tests to ensure components _integrate_ well, and E2E tests that it all works from the user's perspective.

BTW, as you mentioned, what we're currently calling integration tests are closer to E2E tests, just without actually running the k6 binary. So it's even a worse idea to suggest we should have a lot of them, or worse, prefer them over lower level tests.

I see your point about middle-layer tests (i.e. integration tests) being troublesome, but just as frequent refactoring is a pain-point for unit tests, mocking and test setup are a chore for integration tests. So I'm against trying to rewrite these as even higher-level almost-E2E tests to avoid the mocking, but try to make mockable APIs to begin with, and test helpers as we do now to assist with the setup boilerplate.

I think we fundamentally disagree on testing strategies, but a) we won't make much progress discussing it here, and b) I'd like to involve more people in this discussion. So let's schedule it for tomorrow's sync?

Reply:
Yes, we apparently have different preferences here, and :+1: for discussing it more with the rest of the team. 

Until then, can you remove the opinion parts of this we disagree on, so only the more factual bits on which we have consensus remain? This will allow us to merge the PR more quickly, which is important, since it's blocking a few things.

```suggestion
// but also very expensive to run. 
```